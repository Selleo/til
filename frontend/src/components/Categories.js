import React, { useEffect } from 'react'
import { useSelector } from 'react-redux'
import { NavLink } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { saveCategoryPosts } from '../store/actions/actions'
import Icon from './UI/Icon'
import { sortCategories } from '../utils/array/helpers.js'
import { toast } from 'react-toastify'
import { useIsOnRoute } from '../utils/customHooks/useIsOnRoute'

const Categories = () => {
  const categories = useSelector(state => state.categories)
  const dispatch = useDispatch()
  const disable = useIsOnRoute(['add', 'edit'])

  useEffect(() => {
    const currentCategory = document.querySelector(
      '.categories__single-category.-active'
    )

    if (currentCategory) {
      currentCategory.scrollIntoView({
        behavior: 'smooth',
      })
    }
  })

  const blockClickCategory = e => {
    e.preventDefault()
    toast("Can't change view category while post is creating/editing")
  }

  const handleClick = (e, id) => {
    disable && blockClickCategory(e)
    dispatch(saveCategoryPosts(id))
  }

  const sortedCategories = sortCategories(categories)

  return sortedCategories.map(({ id, name }) => (
    <NavLink
      key={name}
      to={`/category/${name}`}
      className="categories__single-category"
      activeClassName="-active"
    >
      <div className="categories__icon">
        <Icon categoryName={name} />
      </div>
      <div className="categories__name">
        <div className="categories__name" onClick={e => handleClick(e, id)}>
          {name}
        </div>
      </div>
    </NavLink>
  ))
}

export default Categories
