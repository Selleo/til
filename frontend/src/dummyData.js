export const initialState = {
  categories: [],
  currentUser: null,
  users: [],
  posts: []
};

export const user = {
  email: "l.walczak@selleo.com",
  firstName: null,
  image:
    "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
  lastName: null,
  uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
};

export const allUsers = [
  {
    email: "l.walczak@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
  },
  {
    email: "maciekPlacek@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAIdsdfjdkjdfi/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-dadlfj33343443"
  }
];

export const categories = [
  { id: 1, name: "java" },
  { id: 2, name: "javascript" }
];

export const posts = [
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adsfdf",
    categoriesIds: [2, 1],
    id: 1,
    isPublic: false,
    likes: [
      {
        post_id: 1,
        user_uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
      }
    ],
    likesCount: 1,
    title: "title "
  },
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adfefsddf",
    categoriesIds: [3],
    id: 2,
    isPublic: false,
    likes: [],
    likesCount: 0,
    title: "maciekdfd"
  }
];
