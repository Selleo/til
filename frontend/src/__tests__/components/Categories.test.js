import React from "react";
import { Provider } from "react-redux";
import { render, cleanup } from "@testing-library/react";
import configureMockStore from "redux-mock-store";
import Categories from "../../components/Categories";
import { categories } from "../../dummyData";

describe("<Categories/>", () => {
  const mockStore = configureMockStore([]);
  let store;

  beforeEach(() => {
    store = mockStore({
      categories,
    });
  });

  afterEach(cleanup);

  it("renders properly", () => {
    store = mockStore({
      categories,
    });
    const { asFragment } = render(
      <Provider store={store}>
        <Categories />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("when there is no categories provided", () => {
    beforeEach(() => {
      store = mockStore({
        categories: [],
      });
    });

    it("renders empty list", () => {
      const { queryAllByTestId } = render(
        <Provider store={store}>
          <Categories />
        </Provider>
      );
      const items = queryAllByTestId("category-item");

      expect(items.length).toEqual(0);
    });
  });
});
