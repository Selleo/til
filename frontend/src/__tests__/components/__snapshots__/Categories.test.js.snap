// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Categories/> renders properly 1`] = `
<DocumentFragment>
  <p>
    java
  </p>
  <p>
    javascript
  </p>
</DocumentFragment>
`;
